### DISCUSSION TITLE
topic_example_1

### DISCUSSION THREAD

**Discussion by minsuLee-kr** [Road-Balance/gcamp_ros2_basic#59] [Q&A] [answered]
topic_example_1코드입니다.

class TwistPubNode(Node):

    def __init__(self):
        super().__init__('twist_pub_node')
        self.get_logger().info(f'TwistPubNode Created at {self.get_clock().now().to_msg().sec}' )
        self.twist_publisher = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
        self.create_timer(0.2, self.timer_callback)

    def timer_callback(self):
        msg = Twist()
        msg.linear.x = random.random()
        msg.angular.z = random.uniform(-1.0, 1.0)
        self.get_logger().info(
            f'Linear X velocity : {msg.linear.x} / Angular Z velocity : {msg.angular.z}')
        self.twist_publisher.publish(msg)

    def main(args=None):
        rclpy.init(args=args)
       twist_pub_node = TwistPubNode()
       rclpy.spin(twist_pub_node)
        twist_pub_node.destroy_node()
        rclpy.shutdown()

이 코드에서
rclpy.spin(twist_pub_node) 가 이해가 안됩니다.

twist_pub_node = TwistPubNode()에서
클래스인 TwistPubNode() 가 인스턴스인 twist_pub_node 에 대입?되었으니 __init__함수가 실행되고,

rclpy.spin(twist_pub_node)에서
twist_pub_node라는 node를 반복하므로 

      def __init__(self):
              super().__init__('twist_pub_node')
              self.get_logger().info(f'TwistPubNode Created at {self.get_clock().now().to_msg().sec}' )
              self.twist_publisher = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
              self.create_timer(0.2, self.timer_callback)
가 계속 반복된다고 생각했는데 터미널의 [INFO]를 보니 그렇지 않은것 같고...spin이정확히 무엇을 반복시키는지 알고싶습니다.
https://docs.ros2.org/latest/api/rclpy/api/init_shutdown.html
에서는 Callbacks이 실행된다고 하던데 의미를 잘 모르겠습니다.
어떤함수가 어떻게 반복되는지 자세히 알고싶습니다.

p.s.  spin이 __init__함수 전체를 반복하지 않는다고 생각한 이유는 이하입니다.

(ROS 2 foxy) dlalstn0079@DESKTOP-6LEK1T5:~/ros2_ws/src/gcamp_ros2_basic/basic_topic_pkg/basic_topic_pkg$ python3 topic_example_1_publisher.py
[INFO] [1658252777.983990600] [twist_pub_node]: TwistPubNode Created at 1658252777
[INFO] [1658252778.186700400] [twist_pub_node]: Linear X velocity : 0.7329105604803374 / Angular Z velocity : 0.8459866642104961
[INFO] [1658252778.387299600] [twist_pub_node]: Linear X velocity : 0.4749425874640242 / Angular Z velocity : 0.995501985191737
[INFO] [1658252778.587302300] [twist_pub_node]: Linear X velocity : 0.7574395912253508 / Angular Z velocity : 0.1484467444177311

      self.get_logger().info(f'TwistPubNode Created at {self.get_clock().now().to_msg().sec}' )
코드의 출력인
[INFO] [1658252777.983990600] [twist_pub_node]: TwistPubNode Created at 1658252777
가 한번만 출력됩니다.

**Comment by kimsooyoung** [comment, marked as answer]
좋은 질문입니다. 

우선, `__init__`은 ros 2 와는 무관한, 파이썬 문법으로 **클래스 생성자**라고 불리는 함수입니다. 
우리가 게임 캐릭터를 생성할 때 초기 능력치나, 이름, 성별 등을 지정하지요? 이와 비슷하게 생성자는 처음 클래스(TwistPubNode)를 통해 인스턴스(twist_pub_node)가 생성될 때 **단 한번만 실행되는 함수**랍니다.

클래스 생성 이후에도 계속 무언가 하고 싶은데... 싶으시죠? 이런 작업은 `timer_callback`에서 구현해야 할 것입니다.
더불어 `twist_pub_node`를 `rclpy.spin()`에 전달해주게 되면, rclpy는 twist_pub_node를 계속 주시하면서 주어진 0.2초가 지나갈 때마다 timer_callback 함수를 실행하게 됩니다. 

여기서 callback이란, 마치 파블로프의 개처럼 **특정한 조건이 만족되면 실행되는 함수**를 뜻합니다. 

충분한 답변이 되었을지 모르겠는데요. 추가 질문은 또 남겨주세요! 

열심히 하시는 모습이 보기 좋습니다. 💪💪

**Reply by minsuLee-kr** [reply]
즉 __init__함수는  twist_pub_node = TwistPubNode() 에서 1번만 실행되고,
rclpy.spin(twist_pub_node)에서 __init__함수의 self.create_timer(0.2, self.timer_callback) 가 아닌 timer_callback(self) (node 안의 모든 메서드)가 반복된다고 생각하면 됩니까?

그리고 0.2초마다 시행되는건
      def __init__(self):
          super().__init__('twist_pub_node')
          self.get_logger().info(f'TwistPubNode Created at {self.get_clock().now().to_msg().sec}' )
          self.twist_publisher = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
          self.create_timer(0.2, self.timer_callback)

에서 self.create_timer(0.2, self.timer_callback)를 통해 노드에 이미 0.2초마다 넘겨라고 전달했기때문이라고 생각하면 됩니까?

**Reply by kimsooyoung** [reply]
네, 정확합니다!
